src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.Analysis.cs

```diff
@@ -0,0 +1,139 @@
+                _context.RegisterCompilationStartAction(
+                    c => analyze(c.Compilation, _this));
+            }
+
+            public void RegisterCodeBlockAction(Action<CodeBlockAnalysisContext, CodeStyleOption<TOptionKind>> analyze)
+            {
+                var provider = _codeStyleProvider;
+                _context.RegisterCodeBlockAction(
+                    c => AnalyzeIfEnabled(provider, c, analyze, c.Options, c.SemanticModel.SyntaxTree, c.CancellationToken));
+            }
+
+            public void RegisterSemanticModelAction(Action<SemanticModelAnalysisContext, CodeStyleOption<TOptionKind>> analyze)
+            {
+                var provider = _codeStyleProvider;
+                _context.RegisterSemanticModelAction(
+                    c => AnalyzeIfEnabled(provider, c, analyze, c.Options, c.SemanticModel.SyntaxTree, c.CancellationToken));
+            }
+
+            public void RegisterSyntaxTreeAction(Action<SyntaxTreeAnalysisContext, CodeStyleOption<TOptionKind>> analyze)
+            {
+                var provider = _codeStyleProvider;
+                _context.RegisterSyntaxTreeAction(
+                    c => AnalyzeIfEnabled(provider, c, analyze, c.Options, c.Tree, c.CancellationToken));
+            }
+
+            public void RegisterOperationAction(
+                Action<OperationAnalysisContext, CodeStyleOption<TOptionKind>> analyze,
+                params OperationKind[] operationKinds)
+            {
+                var provider = _codeStyleProvider;
+                _context.RegisterOperationAction(
+                    c => AnalyzeIfEnabled(provider, c, analyze, c.Options, c.Operation.SemanticModel.SyntaxTree, c.CancellationToken),
+                    operationKinds);
+            }
+
+            public void RegisterSyntaxNodeAction<TSyntaxKind>(
+                Action<SyntaxNodeAnalysisContext, CodeStyleOption<TOptionKind>> analyze,
+                params TSyntaxKind[] syntaxKinds) where TSyntaxKind : struct
+            {
+                var provider = _codeStyleProvider;
+                _context.RegisterSyntaxNodeAction(
+                    c => AnalyzeIfEnabled(provider, c, analyze, c.Options, c.SemanticModel.SyntaxTree, c.CancellationToken),
+                    syntaxKinds);
+            }
+
+            private static void AnalyzeIfEnabled<TContext>(
+                TCodeStyleProvider provider, TContext context, Action<TContext, CodeStyleOption<TOptionKind>> analyze,
+                AnalyzerOptions options, SyntaxTree syntaxTree, CancellationToken cancellationToken)
+            {
+                var optionSet = options.GetDocumentOptionSetAsync(syntaxTree, cancellationToken).GetAwaiter().GetResult();
+                if (optionSet == null)
+                {
+                    return;
+                }
+
+                var optionValue = optionSet.GetOption(provider._option);
+                var severity = GetOptionSeverity(optionValue);
```

Comment:
how does this option severity thing work with compiler severity? it seems to get options from GetDocumentOptionAsync, what happen if user set options through rule set?

we let IDE team to do this as a hack before. are we keeping this hack? whether something run or not should be engines decision not each analyzer.

if IDE wants to create its own system, it should build its own system rather than breaking contracts/assumptions on the existing system created for compiler's diagnostic analyzers.

Reply:
I believe the ultimately, severity is not something analyzer author will use since they can't know what is the last effective severity will be that went through all error mapping such as rule set, warnaserror and etc.

but it is the engine which decide what to run and when based one severity such as hidden, none or refactoring only.

Reply:
> we let IDE team to do this as a hack before. 

I don't know what this means.  IDE didn't do anything here.  This is the API the analyzer system provided to support the options system the IDE has had since some of hte earliest days of Roslyn.  If you want another system used, can you please provide such a system?  IDE is simply using what provided and recommended that it use *by* the analyzer team.

> if IDE wants to create its own system, it should build its own system rather than breaking contracts/assumptions on the existing system created for compiler's diagnostic analyzers.

The diagnostic analyzer system was built *from day one* to be a system that needed to be suitable for IDE scenarios.  The APIs (including teh options system) was provided by the analyzer system.  The IDE never created its own system.  It used what was provided.  Note: the analyzer subsytem was provided by teh analyzer team *and* many features were ported over to use it by that team.  If this system is not sufficient, then there's nothing the IDE can do about that.  We depend on the APIs supplied by the analyzer component.

> but it is the engine which decide what to run and when based one severity such as hidden, none or refactoring only.

Great.  The moment the engine can provide that information, we can consume it on the IDE side.  Until then, i have no idea what hte IDE should do since this is the *only* API available to do this computation.

Reply:
severity is not something analyzer should be able to play with. but IDE team wanted it since they didn't want to use ruleset editor already provided but provide option page in IDE which we have been telling third part analyzer not to do.

so that is what I meant by hack.

...

sam is working on the right way to handle this. which takes time. 

..

> The diagnostic analyzer system was built from day one to be a system that needed to be suitable for IDE scenarios. The APIs (including teh options system) was provided by the analyzer system.

no. diagnostic analyzer is built from day one for compiler. we had some thing for IDE which we throw away for new diagnostic analyzer built for compiler. that's what I mean by creating IDE own stuff if diagnostic analyzer built for compiler is not flexible enough to support everything IDE wants.

since compiler doesn't work the way IDE work, it let something that is hard for IDE to support such as compiler end analyzer in good perf and also, in other cases, it doesn't support things IDE want to support such as solution wide analyzer.

to let system built for compiler to work the way IDE wants, we introduced hacks including analyzer itself checking severity and change its behavior.